ext/formatting_v2.go

```diff
@@ -0,0 +1,772 @@
+		}
+		af.buf = append(af.buf, argStr...)
+		return nil
+	case types.DurationType:
+		argDur, ok := arg.Value().(time.Duration)
+		if !ok {
+			return fmt.Errorf("type conversion error from '%s' to '%s'", arg.Type(), types.DurationType)
+		}
+		af.buf = strconv.AppendFloat(af.buf, argDur.Seconds(), 'f', -1, 64)
+		af.buf = append(af.buf, "s"...)
+		return nil
+	case types.TimestampType:
+		argTime, ok := arg.Value().(time.Time)
+		if !ok {
+			return fmt.Errorf("type conversion error from '%s' to '%s'", arg.Type(), types.TimestampType)
+		}
+		af.buf = argTime.UTC().AppendFormat(af.buf, time.RFC3339Nano)
+		return nil
+	case types.NullType:
+		af.buf = append(af.buf, "null"...)
+		return nil
+	case types.TypeType:
+		argType, ok := arg.Value().(string)
+		if !ok {
+			return fmt.Errorf("type conversion error from '%s' to '%s'", arg.Type(), types.TypeType)
+		}
+		af.buf = append(af.buf, argType...)
+		return nil
+	case types.ListType:
+		argList, ok := arg.(traits.Lister)
+		if !ok {
+			return fmt.Errorf("type conversion error from '%s' to '%s'", arg.Type(), types.ListType)
+		}
+		argIter := argList.Iterator()
+		af.buf = append(af.buf, "["...)
+		if argIter.HasNext() == types.True {
+			if err := af.format(argIter.Next()); err != nil {
+				return err
+			}
+			for argIter.HasNext() == types.True {
+				af.buf = append(af.buf, ", "...)
+				if err := af.format(argIter.Next()); err != nil {
+					return err
+				}
+			}
+		}
+		af.buf = append(af.buf, "]"...)
+		return nil
+	case types.MapType:
+		argMap, ok := arg.(traits.Mapper)
+		if !ok {
+			return fmt.Errorf("type conversion error from '%s' to '%s'", arg.Type(), types.MapType)
+		}
+		argIter := argMap.Iterator()
+		ents := []formattedMapEntryV2{}
+		for argIter.HasNext() == types.True {
+			key := argIter.Next()
+			val, ok := argMap.Find(key)
+			if !ok {
+				return fmt.Errorf("key missing from map: '%s'", key)
+			}
+			keyStr, err := formatStringV2(key)
+			if err != nil {
+				return err
+			}
+			valStr, err := formatStringV2(val)
+			if err != nil {
+				return err
+			}
+			ents = append(ents, formattedMapEntryV2{keyStr, valStr})
+		}
+		sort.SliceStable(ents, func(x, y int) bool {
+			return ents[x].key < ents[y].key
+		})
+		af.buf = append(af.buf, "{"...)
+		for i, e := range ents {
+			if i > 0 {
+				af.buf = append(af.buf, ", "...)
+			}
+			af.buf = append(af.buf, e.key...)
+			af.buf = append(af.buf, ": "...)
+			af.buf = append(af.buf, e.val...)
+		}
+		af.buf = append(af.buf, "}"...)
+		return nil
+	default:
+		return stringFormatErrorV2(runtimeID, arg.Type().TypeName())
+	}
+}
+
+func formatStringV2(arg ref.Val) (string, error) {
+	var fmter appendingFormatterV2
+	if err := fmter.format(arg); err != nil {
+		return "", err
+	}
+	return string(fmter.buf), nil
+}
+
+type stringFormatterV2 struct{}
+
+func (c *stringFormatterV2) String(arg ref.Val) (string, error) {
+	return formatStringV2(arg)
+}
+
+func (c *stringFormatterV2) Decimal(arg ref.Val) (string, error) {
```

Comment:
Could you propagate the comments from the v1 file into v2?

Reply:
Which ones? I thought I copy pasted pretty much everything?

Reply:
Oh, just the capitalized methods. They're all exposed, so `golint` will complain if they're not commented.

Reply:
I thought it only did that for public types? Because formatting.go does have them commented either.